com.unity.ml-agents.extensions/Runtime/Sensors/GridSensorComponent.cs

```diff
@@ -202,68 +194,125 @@ public int ObservationStacks
             set { m_ObservationStacks = value; }
         }
 
+        [HideInInspector, SerializeField]
+        internal bool m_UseOneHotTag = true;
+        /// <summary>
+        /// Whether to use one-hot detected tag as observation.
+        /// Note that changing this after the sensor is created has no effect.
+        /// </summary>
+        public bool UseOneHotTag
+        {
+            get { return m_UseOneHotTag; }
+            set { m_UseOneHotTag = value; }
+        }
+
+        [HideInInspector, SerializeField]
+        internal bool m_CountColliders = false;
+        /// <summary>
+        /// Whether to use the number of count for each detectable tag as observation.
+        /// Note that changing this after the sensor is created has no effect.
+        /// </summary>
+        public bool CountColliders
+        {
+            get { return m_CountColliders; }
+            set { m_CountColliders = value; }
+        }
+
         /// <inheritdoc/>
         public override ISensor[] CreateSensors()
         {
-            m_Sensor = new GridSensor(
-                m_SensorName,
+            List<ISensor> m_Sensors = new List<ISensor>();
+            m_BoxOverlapChecker = new BoxOverlapChecker(
                 m_CellScale,
                 m_GridSize,
                 m_RotateWithAgent,
-                m_ChannelDepths,
-                m_DetectableObjects,
                 m_ColliderMask,
-                m_DepthType,
                 gameObject,
-                m_CompressionType,
-                m_MaxColliderBufferSize,
-                m_InitialColliderBufferSize
+                m_DetectableObjects,
+                m_InitialColliderBufferSize,
+                m_MaxColliderBufferSize
             );
+            // debug data is positive int value and will trigger data validation exception if SensorCompressionType is not None.
+            m_DebugSensor = new GridSensorBase("DebugGridSensor", m_CellScale, m_GridSize, new int[] { 1 }, m_DetectableObjects, SensorCompressionType.None);
```

Comment:
Maybe lazily allocate this in OnDrawGizmos()?

Reply:
That will cause several issues. When there's any changes on component, the debug sensor has to be updated too. So if we turn on gizmo -> turn it off -> change some config -> and turn it back on, it will not be updated. 

Also decoupling the creation of DebugSensor and BoxOverlapChecker cause errors in editor. m_DebugSensor will attach to the existing BoxOverlapChecker at creation time, but editor refresh will keep re-creating new BoxOverlapChecker and the debug sensor won't know about it. 

I think in terms of performance it's not causing too much trouble since when showGizmo=false it doesn't trigger any unnecessary calls.